MonoGame.Framework/Audio/DynamicSoundEffectInstance.OpenAL.cs

```diff
@@ -0,0 +1,189 @@
+
+            DynamicSoundEffectInstanceManager.AddInstance(this);
+        }
+
+        private void PlatformPause()
+        {
+            AL.GetError();
+            AL.SourcePause(SourceId);
+            ALHelper.CheckError("Failed to pause the source.");
+        }
+
+        private void PlatformResume()
+        {
+            AL.GetError();
+            AL.SourcePlay(SourceId);
+            ALHelper.CheckError("Failed to play the source.");
+        }
+
+        private void PlatformStop()
+        {
+            DynamicSoundEffectInstanceManager.RemoveInstance(this);
+
+            AL.GetError();
+            AL.SourceStop(SourceId);
+            ALHelper.CheckError("Failed to stop the source.");
+
+            // Remove all queued buffers
+            AL.BindBufferToSource(SourceId, 0);
+            while (_queuedBuffers.Count > 0)
+            {
+                var buffer = _queuedBuffers.Dequeue();
+                buffer.Dispose();
+            }
+        }
+
+        private void PlatformSubmitBuffer(byte[] buffer, int offset, int count)
+        {
+            // Get a buffer
+            OALSoundBuffer oalBuffer = new OALSoundBuffer();
+
+            // Bind the data
+            if (offset == 0)
+            {
+                oalBuffer.BindDataBuffer(buffer, _format, count, _sampleRate);
+            }
+            else
+            {
+                // BindDataBuffer does not support offset
+                var offsetBuffer = new byte[count];
+                Array.Copy(buffer, offset, offsetBuffer, 0, count);
+                oalBuffer.BindDataBuffer(offsetBuffer, _format, count, _sampleRate);
+            }
+
+            // Queue the buffer
+            AL.SourceQueueBuffer(SourceId, oalBuffer.OpenALDataBuffer);
+            ALHelper.CheckError();
+            _queuedBuffers.Enqueue(oalBuffer);
+
+            // If the source has run out of buffers, restart it
+            if (_state == SoundState.Playing)
+            {
+                AL.SourcePlay(SourceId);
+                ALHelper.CheckError("Failed to resume source playback.");
+            }
+        }
+
+        private void PlatformDispose(bool disposing)
+        {
+            // Stop the source and bind null buffer so that it can be recycled
+            AL.GetError();
+            if (AL.IsSource(SourceId))
+            {
+                AL.SourceStop(SourceId);
+                AL.BindBufferToSource(SourceId, 0);
+                ALHelper.CheckError("Failed to stop the source.");
+                controller.RecycleSource(SourceId);
+            }
+            
+            if (disposing)
+            {
+                while (_queuedBuffers.Count > 0)
+                {
+                    var buffer = _queuedBuffers.Dequeue();
+                    buffer.Dispose();
+                }
+
+                DynamicSoundEffectInstanceManager.RemoveInstance(this);
+            }
+        }
+
+        internal void UpdateQueue()
```

Comment:
I guess this is all because OpenAL has no event fired when a buffer is finished processing?

In XAudio2 it has a callback... but it occurs from some internal XAudio2 thread correct?

So I guess at the moment these callbacks are occuring one different threads on different platforms.  Ideally we would make this more consistent.